### PR TITLE
SJRK-223: Merge the Server project into the UI project

### PR DESCRIPTION
**Pull request by BlueSlug**
This update merges this repository with the Server one, previously found here: https://github.com/fluid-project/sjrk-story-telling-server

The commit history of the latter is preserved, though some of the individual file history will be lost. E.g. for the Dockerfile, since we're using the server's copy, I've kept the server Dockerfile's history.

Naturally, this update requires no changes to the server repo :)

### CODE REVIEW

**Comment by cindyli** [comment]
@BlueSlug, do you have problem running UI tests? File paths in html files for tests seem not adjusted to the new source code file structure. See [one example](https://github.com/BlueSlug/sjrk-story-telling/blob/SJRK-223/tests/ui/html/dynamicViewComponentManager-Tests.html#L5-L14).

The error I ran into even when hosting "sjrk-story-telling" as the root directory of the web server:
screen shot 2019-03-08 at 9 22 54 am

**Comment by BlueSlug** [comment]
Sorry about that @cindyli, I should have added a note that the static tests are now meant to be run via the Node server (`node ./index.js`), at least until I get the paths set up in a more flexible way. Do you think this is okay?

**Comment by cindyli** [comment]
Thanks for the clarification, @BlueSlug. A note for now sounds a good idea.

File: README.md [hunk 1]
```diff
@@ -1,61 +1,110 @@
...
-
-### Installing
-Run `npm install` to install all of the package's dependencies.
-
-### Running
-You may use any webserver software that can run Node.js packages. E.g. using [Browsersync](https://www.browsersync.io/) via the command `browser-sync start -s -f "**"`, the site will be available at `http://localhost:3000/`.
-
-### Testing
-Testing is provided by [jqUnit](https://docs.fluidproject.org/infusion/development/jqUnit.html). If the site is being run at `http://localhost:3000/`, navigate to `http://localhost:3000/tests/all-tests.html` to see the results of all tests and see further details of individual tests.
+# SJRK Story Telling Server
```
Comment (cindyli) [inline README.md]: Could you add a section in README to explain how to run tests, including node tests and browser tests? Thanks.

File: README.md [hunk 2]
```diff
@@ -1,61 +1,110 @@
...
-    - if the folder structure for the new theme is different from the original location of these files, you could update the `pageSetup.resourcePrefix` option of each new page by putting it in the new base grade
-- Add any custom assets required, such as new images, sounds, or fonts.
-- Create a common CSS file with all of the rules for the new theme
-- Add a link in each new HTML file for any applicable new CSS or JavaScript files
-- For each new HTML file, adjust the page component's creation function to use the new custom component
-    - e.g. update `sjrk.storyTelling.page.storyEdit` to `sjrk.storyTelling.myCustomTheme.storyEdit`
-
-### Licenses
-The SJRK Storytelling Tool is provided under either the New BSD license or the Educational Community License, Version 2.0. Please see LICENSE.txt for more details.
+Custom themes can be added to the Storytelling Tool by a process detailed in the documentation of the UI project. If you have a set of custom theme grades and files that you would like to use, you need to:
```
Comment (cindyli) [inline README.md]: I'm quite confused with customization steps such as the purposes and connections among "integrator's custom theme grades and files" and "templates variable" and "sjrk.storyTelling.server.loadThemedPage() function". I wonder if it's possible to add an "examples" directory to demonstrate an implementation of a customized theme.

File: README.md [hunk 3]
```diff
@@ -1,61 +1,110 @@
...
+### Using docker-compose
+
+Three files are used for the `docker-compose` definitions:
+
+- `docker-compose.yml`, the base configuration file
+- `docker-compose.dev.yml`, the stateless dev configuration
+- `docker-compose.cloud.yml`, the production configuration that persists the DB and binary uploads directories using the Docker [bind mounts](https://docs.docker.com/storage/bind-mounts/) approach.
+
+The Compose configuration defines three containers:
+- `app`: the story-telling-server app itself, built from the project `Dockerfile`
```
Comment (cindyli) [inline README.md]: Shall "the story-telling-server app" be "the story telling app"?

File: tests/utils/storyFabricator.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2018 OCAD University
+Licensed under the New BSD license. You may not use this file except in compliance with this licence.
+You may obtain a copy of the BSD License at
+https://raw.githubusercontent.com/fluid-project/sjrk-story-telling-server/master/LICENSE.txt
```
Comment (cindyli) [inline tests/utils/storyFabricator.js]: There are a number of places that the github "sjrk-story-telling-server" repo is still referred to. Can you please update? Thanks.

File: README.md [hunk 5]
```diff
@@ -1,61 +1,110 @@
...
+- In the part of storyTellingServerUI.js that defines the `templates` variable, add an entry for your new custom theme that contains any DOM elements that are required to inject markup or templates.
+    - e.g.
+        ```javascript
+        myCustomTheme: {
+            view: '<div class="sjrk-pageBody-container sjrk-pageBody-container-oneColumn"> YOUR CUSTOM CONTAINERS GO HERE </div>',
+            edit: '<div class="sjrk-pageBody-container sjrk-pageBody-with-sidebars"> YOUR CUSTOM CONTAINERS GO HERE </div>',
+            browse: '<div class="sjrk-pageBody-container sjrk-pageBody-container-oneColumn"> YOUR CUSTOM CONTAINERS GO HERE</div>'
+        }
+        ```
+- In each of the HTML files in the `ui` directory, update the call to `sjrk.storyTelling.server.loadThemedPage` to use your desired theme
```
Comment (cindyli) [inline README.md]: I noticed `sjrk.storyTelling.server.loadThemedPage` is now renamed to `sjrk.storyTelling.loadThemedPage` (the removal of `server` from the namespace).

**Comment by BlueSlug** [comment]
@cindyli I have made a major revision of the README.md, it should be much better structured and accurate, now

File: README.md (hunk 2 above)
Comment (BlueSlug) [reply README.md]: I have restored much of the old theming instructions which point any readers to the `karisma` or `learningReflections` directories for examples of this process

**Comment by cindyli** [comment]
The blank page and an error I hit at viewing a story or on the page after clicking the "publish" button:

Screen Shot 2019-03-11 at 3 10 55 PM

**Comment by BlueSlug** [comment]
@cindyli the error is a result of the name "sjrk.storyTelling.server" no longer being part of that function's name, and also because that gradename isn't within the scope of the application at the UI level (where this function is called). Adding tests for the functions in this file will take a bit of work. Should I do it as part of this Jira or pass it to another issue?

**Comment by cindyli** [comment]
Merged at 9864a6a8766094ee331c1dcb7921d2869f46d25f

**Comment by BlueSlug** [comment]
@cindyli as we discussed, I have added [SJRK-238](https://issues.fluidproject.org/browse/SJRK-238) to add integration tests for the UI and server code and hopefully prevent this and other potential regressions.

**Pull request merged**